src/vault/client/http.clj

```diff
@@ -0,0 +1,496 @@
+(ns vault.client.http
+  (:require
+    [cheshire.core :as json]
+    [clj-http.client :as http]
+    [clojure.string :as str]
+    [clojure.tools.logging :as log]
+    [clojure.walk :as walk]
+    [com.stuartsierra.component :as component]
+    (vault
+      [core :as vault]
+      [lease :as lease]
+      [timer :as timer])))
+
+
+;; ## API Utilities
+
+(defn- kebabify-keys
+  "Rewrites keyword map keys with underscores changed to dashes."
+  [value]
+  (let [kebab-kw #(-> % name (str/replace "_" "-") keyword)
+        xf-entry (juxt (comp kebab-kw key) val)]
+    (walk/postwalk
+      (fn xf-maps [x]
+        (if (map? x)
+          (into {} (map xf-entry) x)
+          x))
+      value)))
+
+
+(defn- clean-body
+  "Cleans up a response from the Vault API by rewriting some keywords and
+  dropping extraneous information. Note that this changes the `:data` in the
+  response to the original result to preserve accuracy."
+  [response]
+  (->
+    (:body response)
+    (kebabify-keys)
+    (assoc :data (:data (:body response)))
+    (->> (into {} (filter (comp some? val))))))
+
+
+(defn- api-error
+  "Inspects an exception and returns a cleaned-up version if the type is well
+  understood. Otherwise returns the original error."
+  [ex]
+  (let [data (ex-data ex)
+        status (:status data)]
+    (if (and status (<= 400 status))
+      (let [body (try
+                   (json/parse-string (:body data) true)
+                   (catch Exception _
+                     nil))
+            errors (if (:errors body)
+                     (str/join ", " (:errors body))
+                     (pr-str body))]
+        (ex-info (str "Vault API errors: " errors)
+                 {:status status
+                  :errors (:errors body)
+                  :cause ex}))
+      ex)))
+
+
+(defn- do-api-request
+  "Performs a request against the API, following redirects at most twice. The
+  `request-url` should be the full API endpoint."
+  [method request-url req]
+  (let [redirects (::redirects (meta req) 0)]
+    (when (<= 2 redirects)
+      (throw (ex-info (str "Aborting Vault API request after " redirects " redirects")
+                      {:method method, :url request-url})))
+    (let [resp (try
+                 (http/request (assoc req :method method :url request-url))
+                 (catch Exception ex
+                   (throw (api-error ex))))]
+      (if-let [location (and (#{303 307} (:status resp)) (get-in resp [:headers "Location"]))]
+        (do (log/debug "Retrying API request redirected to " location)
+            (recur method location (vary-meta req assoc ::redirects (inc redirects))))
```

Comment:
It seems like you could just `assoc` into the request rather than apply metadata.